src/qibo/models/encodings.py

```diff
@@ -323,6 +324,106 @@ def unary_encoder_random_gaussian(
+            \\, \\sum_{j = 1}^{d} \\, x_{j} \\, \\ket{b_{j}}
+
+    Args:
+        data (ndarray): :math:`1`-dimensional array of data to be loaded.
+        nqubits (int): number of qubits.
+        weight (int): Hamming weight that defines the subspace in which ``data`` will be encoded.
+        full_hwp (bool, optional): if ``False``, includes Pauli-:math:`X` gates that prepare the
+            first bitstring of Hamming weight ``k = weight``. If ``True``, circuit is full
+            Hamming weight preserving. Defaults to ```False``.
+        optimize_controls (bool, optional): if ``True``, removes unnecessary controlled operations.
+            Defaults to ``True``.
+        kwargs (dict, optional): Additional arguments used to initialize a Circuit object.
+            For details, see the documentation of :class:`qibo.models.circuit.Circuit`.
+
+    Returns:
+        :class:`qibo.models.circuit.Circuit`: Circuit that loads ``data``in
+        Hamming-weight-:math:`k` representation.
+
+    References:
+        1. R. M. S. Farias, T. O. Maciel, G. Camilo, R. Lin, S. Ramos-Calderer, and L. Aolita,
+        *Quantum encoder for fixed Hamming-weight subspaces*
+        `arXiv:2405.20408 [quant-ph] <https://arxiv.org/abs/2405.20408>`_.
+    """
+    complex_data = bool(data.dtype in [complex, np.dtype("complex128")])
+
+    initial_string = np.array([1] * weight + [0] * (nqubits - weight))
+    bitstrings, targets_and_controls = _ehrlich_algorithm(initial_string)
+
+    """Calculate all gate phases necessary to encode the amplitudes."""
+    _data = np.abs(data) if complex_data else data
+    thetas = _generate_rbs_angles(_data, nqubits, architecture="diagonal")
+    thetas = np.asarray(thetas, dtype=type(thetas[0]))
+    phis = np.zeros(len(thetas) + 1)
+    if complex_data:
+        phis[0] = _angle_mod_two_pi(-np.angle(data[0]))
+        for k in range(1, len(phis)):
+            phis[k] = _angle_mod_two_pi(-np.angle(data[k]) + np.sum(phis[:k]))
+
+    last_qubit = nqubits - 1
+
+    circuit = Circuit(nqubits, **kwargs)
+    if not full_hwp:
+        circuit.add(
+            gates.X(qubit) for qubit in range(last_qubit, last_qubit - weight, -1)
+        )
+
+    if optimize_controls:
+        controls_to_remove = list(range(last_qubit, last_qubit - weight, -1))
+        indices = [
+            int(binom(nqubits - j, weight - j)) - 1 for j in range(weight - 1, 0, -1)
+        ]
+
+    for k, ((targets, controls), theta, phi) in enumerate(
+        zip(targets_and_controls, thetas, phis)
+    ):
+        targets = list(last_qubit - np.asarray(targets))
+        controls = list(last_qubit - np.asarray(controls))
+        controls.sort()
+
+        if optimize_controls:
+            comparison = k < np.asarray(indices)
+            controls_to_remove = list(np.argwhere(comparison == True).flatten())
+            for control_index in controls_to_remove[::-1]:
+                controls.pop(control_index)
```

Comment:
I am not completely sure about this, but can't you simplify this by just taking
```suggestion
            controls = controls[k > np.asarray(indices)]
```
or something along these lines?